[yongkheehou] iP

yongkheehou opened the pull request:
### DailyTasks

> We often overestimate how often we do things, but tracking eliminates this weakness. When we track we just have the facts, not our subjective reflections. - [Betsy Ramser Jaime](https://medium.com/@betsyjaime)

With the DailyTasks app, you can
- save all the tasks you need to do in a day
- have an easy-to-use GUI to list all your tasks
- mark and unmark your tasks as completed

To _use_ it 😄 (It is ~$9.90 a month~ **FREE** of charge!)
1. Download the JAR file in the releases tab on this repo's homepage
2. Save the JAR file to a folder of your choice
3. run `java -jar DailyTasks.jar` in the folder you saved the JAR file

To _contribute_ to the project, make sure
- [x] you are using Java 17
- [x] clone the repo to your computer
- [x] push your commits to a new branch and make a pull request

 Here's the entrypoint of the app in `src/main/java/duke/DailyTasks.java` for reference
```
public static void main(String[] args) {
    launch(args);
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

yongkheehou committed:
complete level 0

yongkheehou committed:
complete level 1

yongkheehou committed:
complete level 2: add + list tasks as strings

yongkheehou committed:
separate separator to its own class

yongkheehou committed:
rename bot

yongkheehou committed:
shift logic for printing tasks out to formatter class

yongkheehou committed:
create task class

yongkheehou committed:
complete level 3: mark as done

yongkheehou committed:
complete level 4: ToDos, Events, Deadlines

yongkheehou committed:
add first test

yongkheehou committed:
add first test

yongkheehou committed:
move tests to another folder and add test2

yongkheehou committed:
complete A-TextUiTesting: added 3 automated tests

yongkheehou committed:
add 2 exception handling

yongkheehou committed:
complete Level-5: adding 2 exceptions

yongkheehou committed:
change array to arraylist

yongkheehou committed:
complete Level-6: delete tasks

yongkheehou committed:
complete A-Enums: add enums for tasks

yongkheehou committed:
abstract the task manager to separate class

yongkheehou committed:
TaskManager.java: add javadoc for TaskManager class

yongkheehou committed:
Complete Level-7: Save

Implement the functionality of automatically saving changes to the
list of tasks in DailyTasks.java

yongkheehou committed:
Merge branch 'branch-Level-7'

Add the automatic saving of list functionality to the chatbot

yongkheehou committed:
DailyTasks.java: abstract out file management to a separate class

yongkheehou committed:
Complete Level-8: Add support for Dates and Times in Deadline

store deadline as LocalDateTime in Deadline.java class
print the dates in an easily readable format

yongkheehou committed:
Event.java: add support for dates and times

implement date time parsing logic for events

yongkheehou committed:
DailyTasks.java: add support for filtering tasks (deadlines/ events) by date and time

yongkheehou committed:
Merge branch 'branch-Level-8'

Implement adding date and time to deadline and events
Add filter command to enable filtering of deadline/ events by date and time

yongkheehou committed:
Refactor to create the main OOP classes

renamed classes to Parser.java, Storage.java, TaskList.java, Ui.java and created Command.java

yongkheehou committed:
Complete A-MoreOOP

Implement Command abstract class and separate individual commands to
its own classes

Modify Parser and Storage classes to use OOP

yongkheehou committed:
Merge branch 'A-MoreOOP'

Implement abstract Command class and separated functionalities to
individual classes

yongkheehou committed:
Format files into packages

create subdirectories and add a general exception

yongkheehou committed:
Format files into packages

create subdirectories and add a general exception

yongkheehou committed:
Merge branch 'A-Packages'

Format the files into packages representing the overarching purpose of
the file

yongkheehou committed:
Merge branch 'add-gradle-support'

Merge the remote branch add-gradle-support to add gradle support to my
iP

yongkheehou committed:
Change application name to duke.DailyTasks in build.gradle

yongkheehou committed:
Fix import orders for checkstyle

yongkheehou committed:
Add javadocs to project to fix gradle build with checkstyle

yongkheehou committed:
Implement tests (date w/o time test is failing) for Parser.java

yongkheehou committed:
Complete A-JUnit

Implement unit tests for InputParser.java
Implement unit tests for DateTimeFormatEnum.java

yongkheehou committed:
Merge branch 'A-JUnit'

Implement unit tests for InputParser.java and DateTimeFormatParser.java

yongkheehou committed:
Complete A-Jar: Package DailyTasks project

Modify build.gradle to package project into jar file

yongkheehou committed:
Merge branch 'A-Jar'

Package the DailyTasks app as a JAR file

yongkheehou committed:
Complete A-JavaDoc

Add JavaDoc for all classes and public methods

yongkheehou committed:
Complete A-CodingStandard

Implement coding standard used in CS2103T

yongkheehou committed:
Complete Level-9: Find

Implement an additional FindTaskCommand to search for tasks

yongkheehou committed:
Merge branch 'A-JavaDoc'

Add JavaDoc to all public classes and methods

yongkheehou committed:
Merge branch 'A-CodingStandard'

Implement CS2103T coding standards in duke repo

yongkheehou committed:
Merge branch 'branch-Level-9'

Implement find command by adding FindTaskCommand class

yongkheehou committed:
Use ui instance to print

Replace system.out.println with ui printer method

yongkheehou committed:
Add FindTaskCommand.java for Level 9

Implement the FindTaskCommand class to represent a find command

yongkheehou committed:
Port over resources from JavaFX tutorial

Add resources folder under src/main/java with css styles, images, and
the fxml for DialogBox and MainWindow

yongkheehou committed:
Port over javafx classes from JavaFX tutorial

Create package duke.javafx to house all the classes involved in
displaying the GUI

yongkheehou committed:
Refactor GUI to work with my current chatbot

Change GUI logic to get response from current Daily Tasks chatbot

yongkheehou committed:
Refactor current commands to return Strings

Change return type of command's execute method to return String instead
of printing inside the execute method and returning void

yongkheehou committed:
Change entrypoint of GUI back to DailyTasks.java

Refactor build.gradle to use duke.dailyTasks as entry pointChange
entrypoint of GUI back to DailyTasks.java

Refactor build.gradle to use duke.dailyTasks as entry pointChange
entrypoint of GUI back to DailyTasks.java

Refactor build.gradle to use duke.dailyTasks as entry pointChange
entrypoint of GUI back to DailyTasks.java

Refactor build.gradle to use duke.dailyTasks as entry pointChange
entrypoint of GUI back to DailyTasks.java

Refactor build.gradle to use duke.dailyTasks as entry pointChange
entrypoint of GUI back to DailyTasks.java

Refactor build.gradle to use duke.dailyTasks as entry pointChange
entrypoint of GUI back to DailyTasks.java

Refactor build.gradle to use duke.dailyTasks as entry pointChange
entrypoint of GUI back to DailyTasks.java

Refactor build.gradle to use duke.dailyTasks as entry pointChange
entrypoint of GUI back to DailyTasks.java

Refactor build.gradle to use duke.dailyTasks as entry point

yongkheehou committed:
Delete unused files from JavaFX tutorial

Delete Duke.java and Main.java which is unused in current project

yongkheehou committed:
Complete Level-10: Implement basic GUI

Follow JavaFX tutorial and ported code over; change all print statements
to return Strings to be displayed on the GUI

yongkheehou committed:
Merge branch 'branch-Level-10'

Add GUI to DailyTasks app and remove the CLI interface

yongkheehou committed:
Complete A-Varargs

Support adding multiple tasks at once by feeding an array of user inputs
to the addTask method

yongkheehou committed:
Merge branch 'A-Varargs'

Support the use of varargs in the addTask method

yongkheehou committed:
GUI Enhancements to UI

Add placeholder for text field and change app title and icon

yongkheehou committed:
Change style for error message

Color DialogBox with green for valid commands and red for invalid ones

yongkheehou committed:
Fix CheckStyle

Remove unused import in DialogBox.java and split long line in
MainWindow.java

yongkheehou committed:
Modify build.gradle to create fat JAR file

Add dependency for JavaFX in build.gradle

ZShunRen commented on src/main/java/duke/tasks/TaskList.java:
```diff
@@ -0,0 +1,173 @@
+     * @return the {@code Task} that was removed.
+     * @throws IndexOutOfBoundsException if the index is out of the range of the task list.
+     */
+    public Task deleteTask(int index) {
```
I think this exception should be caught, otherwise, an invalid index specified by the user will crash your whole program

ZShunRen commented on src/main/java/duke/tasks/TaskList.java:
```diff
@@ -0,0 +1,173 @@
+     * @throws UnknownMessageException if the task type is unknown or unrecognized.
+     * @throws InvalidTodoDescriptionException if the task is a ToDo but lacks a description.
+     */
+    public String addTask(String... userInputs) throws UnknownMessageException, InvalidTodoDescriptionException {
```
I really like the different types of exceptions you have created, which are self-descriptive

ZShunRen commented on src/test/java/duke/parser/DateTimeFormatEnumTest.java:
```diff
@@ -0,0 +1,74 @@
+    @Test
+    public void parse_validDateTimeWithTime_correctLocalDateTime() {
+        String input = "2/12/2019 1800";
+        LocalDateTime expected = LocalDateTime.of(2019, 12, 2, 18, 0);
+
+        Optional<LocalDateTime> result = DateTimeFormatEnum.parse(input);
+
+        assertTrue(result.isPresent(), "Expected a valid LocalDateTime object.");
+        assertEquals(expected, result.get(), "Parsed LocalDateTime should match the expected value.");
```
Nice adherence to separating logical units within a block!

ZShunRen commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,121 @@
+     * @return The formatted message as a string.
+     */
+    public static String formatAddTask(int taskCount, Task task) {
+        return (
+                "Got it. I've added this task:" + "\n" + task.toString() + "\n"
+                        + "Now you have " + taskCount + " tasks in the list." + "\n"
+            );
+    }
```
Not sure why you decided to wrap the return statement this way, but it is still highly readable, which is great!

ZShunRen commented on src/main/java/duke/commands/DeleteTaskCommand.java:
```diff
@@ -0,0 +1,51 @@
+     * @param taskIndex The index of the task to be deleted. A value of -1 indicates an invalid index.
+     */
+    public DeleteTaskCommand(int taskIndex) {
+        super();
```
No harm done, but I think there's no need for this call to the parent constructor

ZShunRen commented on src/main/java/duke/parser/InputParser.java:
```diff
@@ -0,0 +1,97 @@
+     * @throws InvalidInputException if the input format is invalid or cannot be parsed correctly.
+     */
+    public static Command parseUserInput(String userInput) throws InvalidInputException {
+        if (userInput.equals("list")) {
+            return new ListTaskCommand();
+        } else if (userInput.startsWith("unmark")) {
+            return new MarkTaskCommand(false, InputParser.parseTaskIndex(userInput));
+        } else if (userInput.startsWith("mark")) {
+            return new MarkTaskCommand(true, InputParser.parseTaskIndex(userInput));
+        } else if (userInput.startsWith("delete")) {
+            return new DeleteTaskCommand(InputParser.parseTaskIndex(userInput));
+        } else if (userInput.startsWith("filter")) {
+            String dateString = InputParser.parseCommandArgument(userInput, "Invalid date format.");
+            LocalDateTime dateTime = duke.parser.DateTimeFormatEnum.parse(dateString)
+                    .orElseThrow(() -> new InvalidInputException("Invalid date format."));
+            return new FilterTaskCommand(dateTime);
+        } else if (userInput.startsWith("find")) {
+            String keyword = InputParser.parseCommandArgument(userInput, "Invalid find command format.");
+            return new FindTaskCommand(keyword);
+        } else { // we try to add a task (todos/deadline/event)
+            return new AddTaskCommand(userInput);
+        }
+    }
```
I like the simple approach of this way of parsing commands. However, this could be less scalable especially when more commands are needed. Instead I think the commands could have their defining word e.g. "delete" be an attribute of the respective command class.

ZShunRen reviewed (commented):
I liked how clean and organised your code was. There was great adherence to coding standards and very well-done usage of OOP.

thisisaditya17 commented on src/main/java/duke/DailyTasks.java:
```diff
@@ -0,0 +1,66 @@
+import javafx.scene.image.Image;
+import javafx.stage.Stage;
+
+/**
```
Wow!! well done with such a great description

thisisaditya17 commented on src/main/java/duke/commands/Command.java:
```diff
@@ -0,0 +1,31 @@
+ * to define specific behavior.
+ * </p>
+ */
+public abstract class Command {
```
great work creating different classes to handle different tasks!!
Will make your life so easy while testing.
Even I will consider doing so 🙇🏼‍♂️

thisisaditya17 commented on src/main/java/duke/commands/FilterTaskCommand.java:
```diff
@@ -0,0 +1,44 @@
+     * @param dateTime The datetime to filter tasks by.
+     */
+    public FilterTaskCommand(LocalDateTime dateTime) {
+        super();
```
This may be redundant as there's no fields to initialise in Command class? 🤔

thisisaditya17 commented on src/main/java/duke/exceptions/InvalidDeadlineException.java:
```diff
@@ -0,0 +1,25 @@
+package duke.exceptions;
```
great work creating multiple classes to handle multiple exceptions

thisisaditya17 commented on src/main/java/duke/parser/InputParser.java:
```diff
@@ -0,0 +1,97 @@
+     * @return The {@link Command} object representing the action to be taken based on the user input.
+     * @throws InvalidInputException if the input format is invalid or cannot be parsed correctly.
+     */
+    public static Command parseUserInput(String userInput) throws InvalidInputException {
```
maybe try switch statement?

thisisaditya17 commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,163 @@
+     * @return The list of tasks read from the state file.
+     * @throws IOException If an I/O error occurs while reading the file.
+     */
+    public List<Task> loadStateFileToTasksList() throws IOException {
```
Again great abstractions!!

thisisaditya17 commented on src/main/java/duke/tasks/Deadline.java:
```diff
@@ -0,0 +1,67 @@
+     * @return The formatted deadline datetime as a string (e.g., "MMM d yyyy, h:mm a").
+     */
+    public String getBy() {
+        return parsedDateTime.format(DateTimeFormatter.ofPattern("MMM d yyyy, h:mm a"));
```
Maybe create a FORMATTER and add it to task class.

thisisaditya17 reviewed (commented):
Great work!!!
There wasn't much to pick out. Only nitpicks
I learned more from your code. 😆

yongkheehou committed:
Fix build.gradle file

Remove -all in fat JAR file name

yongkheehou committed:
Add assertions

Add assertions to commands, javafx files, DailyTask entrypoint, parser,
and storage classes

yongkheehou committed:
Complete A-Assertions: Merge pull request #1 from yongkheehou/branch-A-Assertions

Add assertions to commands, javafx files, DailyTask entrypoint, parser, and storage classes

Currently there are no checks that the program is executing correctly in the code, other than exceptions to handle expected errors in user input.

This could lead to buggy input not being caught early in the code and thus be propagated down the code.

By writing assertions throughout the code, we can ensure that the code behavior is as we intended through every path the code executes through.

Let's write assertions in the files that handle logic and user input/ output.

yongkheehou committed:
Complete A-CodeQuality

Use switch-case in InputParser.java

yongkheehou committed:
Merge pull request #2 from yongkheehou/branch-A-CodeQuality

Complete A-CodeQuality

yongkheehou committed:
Complete A-Streams

Add basic stream methods in TaskList.java to improve code logic

yongkheehou committed:
Merge pull request #3 from yongkheehou/branch-A-Streams

Complete A-Streams

yongkheehou committed:
Add CI yml file

yongkheehou committed:
Specify SG as locale in github actions yml file

yongkheehou committed:
Specify locale in test file

yongkheehou committed:
Debug with print statements on GH actions

yongkheehou committed:
Add debug information in gradle command

yongkheehou committed:
Specify Locale in format enum

yongkheehou committed:
Remove unused debugging code

yongkheehou committed:
Remove unused imports in test class

yongkheehou committed:
Add ability to set priorities of various tasks

Created an additional SetPriorityCommand class to handle this

yongkheehou committed:
Merge branch 'B-Priority'

Implement logic to set the priority for tasks in our tasklist

yongkheehou committed:
Fix checkstyle errors

Reorder import statements in InputParser.java

yongkheehou committed:
Complete A-UserGuide

Add product demo picture and a preliminary user guide

yongkheehou committed:
Include example outputs in user documentation

yongkheehou committed:
Add hyperlinks to download JAR file

yongkheehou committed:
Add check for invalid indexes

Change the checks for Delete, Mark, SetPriority

yongkheehou committed:
Handle errors for Tasks

Handle purposeful invalid date formats and incomplete tasks

yongkheehou committed:
Fix indexing errors

Modify logic for Delete, Mark, SetPriority

yongkheehou committed:
Fix code quality issues

Address the Javadoc and naming of boolean variables raised by CS2103T
automatic check for code quality

yongkheehou committed:
Update product page and readme of repo

yongkheehou committed:
Fix markdown

yongkheehou committed:
Add sample image of app
